File: LLM_GGUF_graph/graphApp_py/extract_graph.py
from graph import Graph, Node, Tensor
import struct
import os

def print_metadata_as_table(metadata):
    # Lista de prefijos de claves que deseamos extraer
    key_prefixes = ["llama.", "general.", "tokenizer."]
    position = 0  # Para manejar índices

    while position < len(metadata):
        found_key = False  # Para determinar si se encontró un key

        for prefix in key_prefixes:
            start_position = metadata.find(prefix, position)
            if start_position != -1:  # Se encontró un key
                found_key = True

                # Buscar el siguiente key
                end_position = len(metadata)  # Resetear end_position
                for next_prefix in key_prefixes:
                    next_position = metadata.find(next_prefix, start_position + len(prefix))
                    if next_position != -1:
                        end_position = next_position
                        break  # Salir si se encontró el siguiente key

                # Tomar la línea de texto desde start_position hasta end_position
                text_line = metadata[start_position:end_position]
                print(text_line)

                position = end_position - 1  # Continuar desde el final del key encontrado
                break  # Salir del bucle de prefijos si se encontró uno

        # Si no se encontró un key, avanzar la posición
        if not found_key:
            position += 1


def load_gguf(filename):
    graph = {}  # Crear un grafo vacío (o un objeto Graph si tienes una clase Graph definida)
    
    # Variables para los encabezados
    gguf_magic_number = 0
    gguf_version = 0
    tensor_count = 0
    kv_count = 0

    if not os.path.isfile(filename):
        print(f"Error: No se pudo abrir el archivo {filename}")
        return graph  # Retornar un grafo vacío en caso de error

    # Leer el tamaño del archivo completo
    file_size = os.path.getsize(filename)
    print(f"\n >>> Tamaño del archivo:  {file_size}  [bytes]")

    with open(filename, "rb") as file:
        # Leer los primeros datos: GGUF magic number, versión, tensor_count, kv_count
        gguf_magic_number = struct.unpack('I', file.read(4))[0]  # uint32_t
        gguf_version = struct.unpack('I', file.read(4))[0]      # uint32_t
        tensor_count = struct.unpack('Q', file.read(8))[0]      # uint64_t
        kv_count = struct.unpack('Q', file.read(8))[0]          # uint64_t

        # El punto actual en el archivo es donde empieza la metadata
        start = file.tell()

        # Determinar cuántos bytes leer para la metadata
        metadata_bytes_to_read = 5000  # cantidad de metadata a leer
        end = start + metadata_bytes_to_read

        # Verificar que el rango de lectura es válido
        if end > file_size:
            end = file_size

        bytes_to_read = end - start
        if bytes_to_read <= 0:
            print("Error: El rango de lectura no es válido.")
            return graph  # Retornar un grafo vacío

        # Leer la metadata
        metadata_data = file.read(bytes_to_read).decode('utf-8', errors='ignore')

        # Imprimir metadata como tabla
        print(f"\n > gguf_magic_number: {gguf_magic_number}")
        print(f" > gguf_version: {gguf_version}")
        print(f" > tensor_count: {tensor_count}")
        print(f" > kv_count: {kv_count}")

        print_metadata_as_table(metadata_data)

    return graph

File: LLM_GGUF_graph/graphApp_cpp/extract_graph.cpp
#include <iomanip>
#include <vector>
#include <string>
#include <fstream>
#include <iostream>
#include <cstring>
#include "extract_graph.h"

void printMetadataAsTable(const std::string& metadata) {
    // Lista de prefijos de claves que deseamos extraer
    const char* keyPrefixes[] = {"llama.", "general.", "tokenizer."};
    const int numPrefixes = 3; 

    size_t position = 0; // Para manejar índices
    size_t startPosition = 0;
    size_t endPosition = 0;
    std::string textLine;

    while (position < metadata.size()) {
        bool foundKey = false; // Para determinar si se encontró un key

        for (int i = 0; i < numPrefixes; i++) {
            startPosition = metadata.find(keyPrefixes[i], position);
            if (startPosition != std::string::npos) {
                foundKey = true; // Se encontró un key

                // Buscar el siguiente key
                endPosition = std::string::npos; // Resetear endPosition
                for (int j = 0; j < numPrefixes; j++) {
                    endPosition = metadata.find(keyPrefixes[j], startPosition + strlen(keyPrefixes[i]));
                    if (endPosition != std::string::npos) {
                        break; // Salir si se encontró el siguiente key
                    }
                }

                // Si no se encontró un siguiente key, ajustar endPosition al final de la cadena
                if (endPosition == std::string::npos) {
                    endPosition = metadata.size();
                }

                // Tomar la línea de texto desde startPosition hasta endPosition
                textLine = metadata.substr(startPosition, endPosition - startPosition);
                std::cout << textLine << std::endl;

                position = endPosition-1; // Continuar desde el final del key encontrado
                break; // Salir del bucle de prefijos si se encontró uno
            }
        }

        // Si no se encontró un key, avanzar la posición
        if (!foundKey) {
            position++;
        }
    }
}


Graph loadGGUF(const std::string& filename) {
    Graph graph;
    // Variables para los encabezados
    uint32_t gguf_magic_number;
    uint32_t gguf_version;
    uint64_t tensor_count;
    uint64_t kv_count;

    std::ifstream file(filename, std::ios::binary);
    if (!file.is_open()) {
        std::cerr << "Error: No se pudo abrir el archivo " << filename << std::endl;
        return graph;  // Retornar un grafo vacío en caso de error
    }

    // Leer el tamaño del archivo completo
    file.seekg(0, std::ios::end);
    std::streampos fileSize = file.tellg();
    std::cout << "\n >>> Tamaño del archivo:  " << fileSize << "  [bytes]" << std::endl;

    // Ir al principio del archivo
    file.seekg(0, std::ios::beg);

    // Leer los primeros datos: GGUF magic number, versión, tensor_count, metadata_kv_count
    file.read(reinterpret_cast<char*>(&gguf_magic_number), sizeof(gguf_magic_number));
    file.read(reinterpret_cast<char*>(&gguf_version), sizeof(gguf_version));
    file.read(reinterpret_cast<char*>(&tensor_count), sizeof(tensor_count));
    file.read(reinterpret_cast<char*>(&kv_count), sizeof(kv_count));

    // El punto actual en el archivo es donde empieza la metadata
    std::streampos start = file.tellg();

    // Determinar cuántos bytes leer para la metadata
    std::streampos metadataBytesToRead = 5000;  // cantidad de metadata a leer
    std::streampos end = start + metadataBytesToRead;

    // Verificar que el rango de lectura es válido
    if (end > fileSize) {
        end = fileSize;
    }

    std::streampos bytesToRead = end - start;
    if (bytesToRead <= 0) {
        std::cerr << "Error: El rango de lectura no es válido." << std::endl;
        return graph;  // Retornar un grafo vacío
    }

    // Leer la metadata
    std::string metadataData(bytesToRead, '\0');
    file.read(&metadataData[0], bytesToRead);

    // Verificar si la lectura fue exitosa
    if (!file) {
        std::cerr << "Error: Fallo en la lectura del archivo." << std::endl;
        return graph;  // Retornar un grafo vacío
    }

    std::cout << "\n > gguf_magic_number: " << gguf_magic_number << std::endl;
    std::cout << " > gguf_version: " << gguf_version << std::endl;
    std::cout << " > tensor_count: " << tensor_count << std::endl;
    std::cout << " > kv_count: " << kv_count << std::endl;

    // Imprimir metadata como tabla
    printMetadataAsTable(metadataData);

    file.close();
    return graph;
}
